auth-manifest/gen/src/generator.rs

```diff
@@ -0,0 +1,166 @@
+/*++
+
+Licensed under the Apache-2.0 license.
+
+File Name:
+
+   generator.rs
+
+Abstract:
+
+    Caliptra Authorization Manifest generator
+
+--*/
+
+use caliptra_image_gen::ImageGeneratorCrypto;
+use zerocopy::AsBytes;
+
+use crate::*;
+use core::mem::size_of;
+
+/// Authorization Manifest generator
+pub struct AuthManifestGenerator<Crypto: ImageGeneratorCrypto> {
+    crypto: Crypto,
+}
+
+impl<Crypto: ImageGeneratorCrypto> AuthManifestGenerator<Crypto> {
+    /// Create an instance `AuthManifestGenerator`
+    pub fn new(crypto: Crypto) -> Self {
+        Self { crypto }
+    }
+
+    pub fn generate(
+        &self,
+        config: &AuthManifestGeneratorConfig,
+    ) -> anyhow::Result<AuthorizationManifest> {
+        let mut auth_manifest = AuthorizationManifest::default();
+
+        if config.image_metadata_list.len() > AUTH_MANIFEST_IMAGE_METADATA_MAX_COUNT {
+            eprintln!(
+                "Encountered an error converting image_metadata_list, only {} entries supported.",
+                AUTH_MANIFEST_IMAGE_METADATA_MAX_COUNT
+            );
+            return Err(anyhow::anyhow!("Error converting image_metadata_list"));
+        }
+
+        // Generate the Image Metadata List.
+        let slice = config.image_metadata_list.as_slice();
+        auth_manifest.image_metadata_col.image_metadata_array[..slice.len()].copy_from_slice(slice);
+
+        auth_manifest.image_metadata_col.header.entry_count =
+            config.image_metadata_list.len() as u32;
+        auth_manifest.image_metadata_col.header.revision = 0; // [TODO] Need to update this.
+
+        // Generate the preamble.
+        auth_manifest.preamble.marker = AUTH_MANIFEST_MARKER;
+        auth_manifest.preamble.size = size_of::<AuthManifestPreamble>() as u32;
+        auth_manifest.preamble.version = config.version;
+        auth_manifest.preamble.flags = config.flags;
+
+        // Sign the vendor manifest keys.
+        auth_manifest.preamble.vendor_pub_keys = config.vendor_man_key_info.pub_keys;
+
+        let range = AuthManifestPreamble::vendor_signed_data_range();
+
+        let data = auth_manifest
+            .preamble
+            .as_bytes()
+            .get(range.start as usize..)
+            .ok_or_else(|| anyhow::anyhow!("Failed to get vendor signed data range start"))?
+            .get(..range.len())
+            .ok_or(anyhow::anyhow!(
+                "Failed to get vendor signed data range length"
+            ))?;
+
+        let digest = self.crypto.sha384_digest(data)?;
+
+        if let Some(priv_keys) = config.vendor_fw_key_info.priv_keys {
+            let sig = self.crypto.ecdsa384_sign(
+                &digest,
+                &priv_keys.ecc_priv_key,
+                &config.vendor_fw_key_info.pub_keys.ecc_pub_key,
+            )?;
+            auth_manifest.preamble.vendor_pub_keys_signatures.ecc_sig = sig;
+
+            let lms_sig = self.crypto.lms_sign(&digest, &priv_keys.lms_priv_key)?;
```

Comment:
Currently this tool requires the manifest to be signed with both ecdsa and lms keys. Do we want to support manifests that are only signed with ecdsa (since we support this config for Caliptra FW).

Reply:
This is modeled after the Caliptra FW generator which always takes LMS keys as part of the config. I can make the LMS keys optional as part of a separate change.

Reply:
> I can make the LMS keys optional as part of a separate change.

Wouldn't the spec need to change to allow marking LMS keys as optional? Like an extra flag bit or something?

Reply:
What I meant was while creating the authorization manifest, O'll make LMS keys optional and if not provided, won't perform any signing operations with those (obvious). Currently, both Caliptra FW and Auth Manifest generation code expect LMS keys to be provided. During validation of the authorization manifest in Runtime, the value from 'fuse_lms_verify' will be used to validate with the LMS keys. Not specifying LMS keys in the authorization manifest but setting fuse_lms_verify to true should be considered a configuration error.

Reply:
Got it. Updating the tooling in a future PR sounds good.

> During validation of the authorization manifest in Runtime, the value from 'fuse_lms_verify' will be used to validate with the LMS keys.

Can you add this to the md file?

Thanks!

Reply:
Added guidance to the README file. I'll make the LMS keys optional in a follow-up PR.